packages/design-system/src/components/Button/Primitive/ButtonPrimitive.tsx

```diff
@@ -1,78 +1,84 @@
 	variant: T;
 } & P;
 
-export type SharedButtonTypes = {
-	size?: AvailableSizes;
-	icon?: IconName | React.ReactElement;
+export type SharedButtonTypes<S extends AvailableSizes> = {
 	isLoading?: boolean;
 	isDropdown?: boolean;
+	size?: S;
+	icon?: S extends 'S'
+		? IconNameWithSize<'S'>
+		: DeprecatedIconNames | ReactElement | IconNameWithSize<'M'>;
 };
 
-export type BaseButtonProps = Omit<ClickableProps, 'style'> &
-	SharedButtonTypes &
+export type BaseButtonProps<S extends AvailableSizes> = Omit<ClickableProps, 'style'> &
+	SharedButtonTypes<S> &
 	Partial<DataAttributes>;
 
-const ButtonPrimitive = forwardRef(
-	(
-		{
-			className,
-			children,
-			onClick,
-			size = 'M',
-			icon,
-			isLoading = false,
-			isDropdown = false,
-			...props
-		}: BaseButtonProps,
-		ref: Ref<HTMLButtonElement>,
-	) => {
-		return (
-			<Clickable
-				className={classnames(styles.button, className, {
-					[styles['size-S']]: size === 'S',
-				})}
-				{...props}
-				aria-busy={isLoading}
-				ref={ref}
-				onClick={!isLoading ? onClick : () => {}}
-			>
-				<StackHorizontal gap="XS" as="span" align="center">
-					{isLoading && (
-						<span className={styles.button__loading}>
-							<Loading data-test="button.loading" name={icon} aria-hidden />
-						</span>
-					)}
-					{!isLoading && icon && size === 'M' && (
-						<span className={styles.button__icon}>
-							{typeof icon === 'string' ? <Icon name={icon} /> : React.cloneElement(icon, {})}
-						</span>
-					)}
-					{children}
-					{isDropdown && (
-						<span className={styles.button__caret}>
-							<Icon name="talend-caret-down" />
-						</span>
-					)}
-				</StackHorizontal>
-			</Clickable>
-		);
-	},
-);
+function ButtonPrimitiveInner<S extends AvailableSizes>(
+	props: BaseButtonProps<S>,
+	ref?: Ref<HTMLButtonElement>,
+) {
+	const {
+		className,
+		children,
+		onClick,
+		size = 'M',
+		icon,
+		isLoading = false,
+		isDropdown = false,
+		...rest
+	} = props;
+	return (
+		<Clickable
+			className={classnames(styles.button, className, {
+				[styles['size-S']]: size === 'S',
+			})}
+			{...rest}
+			aria-busy={isLoading}
+			ref={ref}
+			onClick={!isLoading ? onClick : () => {}}
+		>
+			<StackHorizontal gap="XS" as="span" align="center">
+				{isLoading && (
+					<span className={styles.button__loading}>
+						<Loading data-test="button.loading" aria-hidden />
+					</span>
+				)}
+				{!isLoading && icon && (
+					<span className={styles.button__icon}>
+						{parseDeprecatedIcon({ iconSrc: icon, size: size === 'S' ? 'S' : 'M' })}
+					</span>
+				)}
+				{children}
+				{isDropdown && (
+					<span className={styles.button__caret}>
+						<SizedIcon size="S" name="chevron-down" />
+					</span>
+				)}
+			</StackHorizontal>
+		</Clickable>
+	);
+}
+
+const ButtonPrimitive = forwardRef(ButtonPrimitiveInner) as <S extends AvailableSizes>(
+	props: BaseButtonProps<S> & { ref?: Ref<HTMLButtonElement> },
+) => ReturnType<typeof ButtonPrimitiveInner>;
```

Comment:
TS needs a little help at this stage. `forwardRef` makes it lose the `S` generic without the typecasting.